<h3 style="color:blue;">Data Design Midterm Submission </h3>
<h4>Section 2 | Shalini Mishra 

## Part 1
Go to http://numbersapi.com and familiarize yourself with the API.
a) Write a program that accesses all trivia results for the numbers from 0 (zero) to 99 using batch requests only (One single query for all the numbers). Print the output result to the screen in the format of [3-digit number with leading zeros] - [TRIVIA] (One number per each line). E.g.,<br>
000 - 0 is the coldest possible temperature old the Kelvin scale.<br>
001 - 1 is the number of dimensions of a line.<br>
002 - 2 is the price in cents per acre the USA bought Alaska from Russia.<br>
...<br>
Note: your output for each number may be different. The website randomizes its trivia responses.!<br>

In [1]:
#importing necessary libraries
import requests as r

#Getting all the trivia results from numbers 0 to 99 as a batch request
url='http://numbersapi.com/0..99'
response=r.get(url)
response.json()

{'0': '0 is the coldest possible temperature old the Kelvin scale.',
 '1': '1 is the number of moons orbiting Earth.',
 '2': '2 is the price in cents per acre the USA bought Alaska from Russia.',
 '3': "3 is the number of witches in William Shakespeare's Macbeth.",
 '4': '4 is the number of nucleobase types in DNA and RNA – adenine, guanine, cytosine, thymine (uracil in RNA).',
 '5': '5 is the number of appendages on most starfish, which exhibit pentamerism.',
 '6': '6 is the standard length (year) of a term in office for a United States senator.',
 '7': '7 is the number of periods, or horizontal rows of elements, in the periodic table.',
 '8': '8 is the number of bits in a byte.',
 '9': '9 is the number of innings in a regulation, non-tied game of baseball.',
 '10': '10 is the Number of dots in a tetractys.',
 '11': '11 is the possible age of the youngest elected pope, Benedict IX.',
 '12': '12 is the number of keys in any standard digital telephone (1 through 9, 0, * and #).',
 '13':

<h5>(c) Parse the JSON string to an internal Python object for further handling. Then wrote code that iterates through the search results and prints the trivia results to the screen.</h5>

In [25]:
import json as js

result=js.loads(response.text)
#Print each of the trivia results
for i in range(0,len(result)):
    #Use format to get the desired outcome
        print('{:03d}'.format(i),'-',result[str(i)])

000 - 0 is the coldest possible temperature old the Kelvin scale.
001 - 1 is the number of moons orbiting Earth.
002 - 2 is the price in cents per acre the USA bought Alaska from Russia.
003 - 3 is the number of witches in William Shakespeare's Macbeth.
004 - 4 is the number of nucleobase types in DNA and RNA – adenine, guanine, cytosine, thymine (uracil in RNA).
005 - 5 is the number of appendages on most starfish, which exhibit pentamerism.
006 - 6 is the standard length (year) of a term in office for a United States senator.
007 - 7 is the number of periods, or horizontal rows of elements, in the periodic table.
008 - 8 is the number of bits in a byte.
009 - 9 is the number of innings in a regulation, non-tied game of baseball.
010 - 10 is the Number of dots in a tetractys.
011 - 11 is the possible age of the youngest elected pope, Benedict IX.
012 - 12 is the number of keys in any standard digital telephone (1 through 9, 0, * and #).
013 - 13 is the Youngest age a minor can rent or

## Part 2: LONGER!
<h5>Web-scraping: Which sellers advertise/sponsor on eBay<br><br></h5>
<h3> 2.a</h3>
<h5>How would you modify your search query to only include buy-it-now (non-auction) items?</h5><br>We will change our URL to "https://www.ebay.com/sch/i.html?_nkw=<b>search term</b>&<b  style="color:blue;">LH_BIN=1</b>"<br>LH_BIN=1 ensures that only buy-it-now options are shown<br><br><b>What is the GET request's variable name corresponding to buy-it-now searches?</b><br>Variable Name:LH_BIN<br><h5>How would you modify your search query to include 100 items per search result page?</h5>Set the variable name,<b>_ipg</b>=100 in the URL.<br>"https://www.ebay.com/sch/i.html?_nkw=<b>search term</b>&<b  style="color:blue;">LH_BIN=1</b>&<b>_ipg</b>=100"<br><h5> What is the GET request's variable name corresponding to items per page searches?</h5><br>Variable name is <b>i_ipg</b>


<h3> 2.b</h3><br>
<h5>Modify your previous code to search for buy-it-now listings of <keyword> and limit the number of items to 100 per page. Use this guide to select the right <keyword> assigned to you<br></h5>
If your last name on Canvas starts with M-R use <keyword>: "iphone 11 pro"<br>
Identify all the sponsored items.<br>
For the first 10 pages of 100 items/page, save all the URLs of sponsored items' pages to the file "sponsored.txt"  and all the URLs of non-sponsored items' pages to the file "non-sponsored.txt" in the same directory as your code. (One URL per line in each file)<br>

In [1021]:
#Importing libraries
import requests 
import re
from requests import get
from bs4 import BeautifulSoup
import time


In [70]:
#creating function
def url_creation(i):
        url = "https://www.ebay.com/sch/i.html?_nkw=iphone+11+pro&_pgn="+str(i)+"&LH_BIN=1&_ipg=100"
        user_agent = {'User-agent': 'Mozilla/5.0'} 
        # Requesting the ebay website access to the server and storing the response
        res = get(url,user_agent)
        # Parsing the response and storing the resulting html
        data = BeautifulSoup(res.text, 'html.parser')
        items = data.select("li.s-item")
        for item in items:
            a = item.find("a")
            if not a:
                continue

            text = item.find_all("h3")
            if re.search(".s-item__title--has-tags",str(text)):  
                #print(a['href'])
                url_clean = re.sub("(.*)\\?.*", r"\1", a['href'])           
                with open("sponsored.txt", "a+") as file:
                    file.write(url_clean+"\n")
                    file.close()
                url_clean=""
            else:
                url_clean = re.sub("(.*)\\?.*", r"\1", a['href'])           
                with open("non-sponsored.txt", "a+") as file:
                    file.write(url_clean+"\n")
                    file.close()
                url_clean=""
                


<i>Setting a timer of 10 seconds between each iteration of the function</i>

In [71]:
import time
for i in range(1,11):
    time.sleep(10)
    url_clean=""
    url_creation(i)

<h3>2.c</h3><br>
<h5>Create two folders in the same directory as your code and name them "sponsored" and "non-sponsored". <br>
Write a program that opens the two files in (b) and downloads each of the pages (URLs) into the folders "sponsored" and "non-sponsored".<br></h5>
Each file should be named as "<item-id>.htm" where you replace "item-id" with the ID of the item you are saving.<br>
E.g., "264616053293.htm" for the item with ID "264616053293".<br>
Note it is always good to put a 2-second pause between queries. Make sure to catch an error and continue if your query runs into problems connecting to eBay (e.g., if your internet is down for 5 seconds, you don't want your entire code to crash).<br>



In [78]:
import os
#function to create folders in current directory
def createFolder(directory):
    try:
        if not os.path.exists(directory):
            os.makedirs(directory)
            print(directory,':Created')
    except OSError:
        print ('Error: Creating directory. ' +  directory)
        

In [79]:
#creating two folders-Sponsored and Non-sponsored
createFolder('./Sponsored/')
createFolder('./Non-Sponsored/')

./Sponsored/ :Created
./Non-Sponsored/ :Created


#### Function for creating files for each product with name as <product_id.htm> on reading the text files (sponsored.txt and non-sponsored.txt)


In [415]:
def item_url(i):
        url=i.rstrip('\n')
        user_agent = {'User-agent': 'Mozilla/5.0'} 
        # Requesting the ebay website access to the server and storing the response
        try:
            res = get(url, user_agent)
        except requests.exceptions.Timeout:
            time.sleep(5)
            res = get(url, user_agent)
        except requests.exceptions.RequestException as e:
            sys.exit(1)                   # bail out of the function  
        #find Item ID using Regex
        pattern=re.search('([0-9]{12})',i)
        #Use the matched pattern for filename
        name=str(pattern.groups()[0])+'.htm'
        with open(name, "wb") as file:
                file.write(res.content) #to include the rich content except the html part
                file.close()

In [416]:
#Opening the file sponsored.text
f = open("sponsored.txt", "r")
#Get the current working direcory: 
org_path=os.getcwd()
#Redirecting to Sponsored Folder
new_path=org_path+"\Sponsored"
#Change the current working direcory: 
os.chdir(new_path)
for x in f:
    time.sleep(5)
    item_url(x)
f.close()


In [477]:
#Getting outside of Sponsored folder
os.chdir('../')

In [478]:
#Opening the file non-sponsored.txt
f = open("non-sponsored.txt", "r")
#Get the current working direcory: 
org_path=os.getcwd()
#Redirecting to Sponsored Folder
new_path=org_path+"\/"+"Non-Sponsored"
#Change the current working direcory: 
os.chdir(new_path)
for x in f:
    time.sleep(5)
    item_url(x)
f.close()


In [179]:
#Getting outside of Non-Sponsored folder
os.chdir('../')
#Reset the directory (Printing to check)
os.getcwd()

'C:\\Users\\shali\\OneDrive\\Documents\\3 WQ Data Design\\Assignments Submitted'

<h3>2.d</h3><br><h5>Write a separate piece of code that loops through the pages you downloaded in (c) and opens and parses them into a Python or Java xxxxsoup-object. Identify and select:</h5><br>
seller name, seller score, item price, # items sold, best offer available, title, returns allowed, shipping price, condition (e.g., used, new, like new, seller refurbished, ...).<br>
In your code, highlight the selector command you choose to obtain each element using comments.


In [556]:
#Creating an empty dataframe with necessary columns to store scraped contents
import pandas as pd
scrape_df=pd.DataFrame(columns=['seller_name', 'seller_score', 'item_price','units_sold','best_offer',
                     'item_title','returns_allowed','shipping_price','item_condition','sponsored'])


In [895]:
#Storing the current directory path
org_path=os.getcwd()
print(org_path)

'C:\\Users\\shali\\OneDrive\\Documents\\3 WQ Data Design\\Assignments Submitted'

#### Function created to scrape through each file in the directory<br>
Passing Parameters: path to the directory, sponsor flag value which is an indicator for items

In [576]:
import numpy as np

def scrape_store(path,sponsor_flag):
    scrape_df=pd.DataFrame(columns=['seller_name', 'seller_score', 'item_price','units_sold','best_offer',
                     'item_title','returns_allowed','shipping_price','item_condition','sponsored'])
    #Iterating through files in a directory
    for file in os.listdir(path):
        #open each file in read bytes mode
        f=open(file, "rb")
        filetext=f.read()
        #parse it through Beautiful soup 
        data = BeautifulSoup(filetext, 'html.parser')

        #Get Seller-name
        seller_name='No info'if data.find('span',class_='mbg-nw') is None else data.find('span',class_='mbg-nw').string

        #store seller score in integer format
        seller_score=np.nan if data.select('span.mbg-l > a') is None else int(data.select('span.mbg-l > a')[0].string)

        #Item Price
        item_price= np.nan if data.find('span',class_='notranslate')is None else data.find('span',class_='notranslate').string
        #,itemprop='price'
        #if NoneType object is found then store as nan(value unavailable)
        units_sold= np.nan if data.find('a',class_='vi-txt-underline') is None else data.find('a',class_='vi-txt-underline').string

        #if we get NoneType object, best offer doesn't exist for the item
        best_offer= True if data.find('div',class_=re.compile('.*boLable')) is not None else False

        #Item title
        title1=data.find('h1',class_='it-ttl',id='itemTitle').get_text()
        #removing unecessary part of the string
        item_title=re.sub('.*\xa0', '', title1)

        #Return Details
        returns_allowed= None if data.find('span',id='vi-ret-accrd-txt') is None else data.find('span',id='vi-ret-accrd-txt').string

        #Shipping Price
        #findinf shipping summary then looking for next as its dynamic across product pages
        shipping= data.find('span',id='shSummary')
        shipping_price=shipping.find_next().get_text()

        #Item Condition
        item_condition=data.find('div',id='vi-itm-cond',itemprop='itemCondition').string

        #Close the file
        f.close()

        #Append to the dataframe
        scrape_df = scrape_df.append({'seller_name': seller_name, 'seller_score': seller_score, 'item_price':item_price,
                                 'units_sold':units_sold,'best_offer':best_offer,'item_title':item_title,
                                 'returns_allowed':returns_allowed,'shipping_price':shipping_price,'item_condition':item_condition,
                                'sponsored':sponsor_flag}, ignore_index=True)
    return(scrape_df)

In [587]:
new_path=org_path+"\Sponsored"
#Change the current working direcory: 
os.chdir(new_path)
#sponsor-flag=1 for sponsored items
sponsored_df=scrape_store(new_path,1)

In [876]:
#Getting outside of Sponsored folder
os.chdir('../')

In [592]:
new_path=org_path+"\/"+"Non-Sponsored"
#Change the current working direcory: 
os.chdir(new_path)
#sponsor-flag=0 for non-sponsored items
nonsponsored_df=scrape_store(new_path,0)

In [547]:
#Getting outside of Non-Sponsored folder
os.chdir('../')

In [588]:
sponsored_df

seller_name seller_score    item_price units_sold best_offer  \
0            the.lux.group          188  US $2,349.00        NaN       True   
1            the.lux.group          188  US $1,799.00     1 sold       True   
2            the.lux.group          188  US $1,999.00     1 sold       True   
3             normabreede0           15  US $1,500.00        NaN       True   
4              klp_customs          148  US $1,799.00        NaN      False   
5              klp_customs          148  US $1,799.00     1 sold      False   
6                  merinsh           84  US $1,700.00     2 sold      False   
7                  dheader        75647    US $699.95        NaN      False   
8       harvardhoney4545_1          184  US $1,699.99        NaN       True   
9                shaneel02           99  US $2,499.99     1 sold       True   
10         swiftphonetechs          127    US $859.00        NaN      False   
11                punkydk0          589    US $719.99        NaN      False   
12                punkydk0          589    US $569.99        NaN      False   
13         swiftphonetechs          127    US $925.00        NaN      False   
14            vombotsupply         5881  US $1,029.99   150 sold      False   
15          the_phone_king        14640  AU $3,047.99        NaN      False   
16          the_phone_king        14640  AU $2,315.99        NaN      False   
17          the_phone_king        14640  AU $2,132.99        NaN      False   
18                be-start         8451  AU $2,210.95        NaN      False   
19                be-start         8451  AU $3,211.55     1 sold      False   
20                be-start         8451  AU $3,029.35        NaN      False   
21            narrowgate15         4934     US $10.99     8 sold      False   
22            narrowgate15         4934      US $8.99     2 sold       True   
23            narrowgate15         4934      US $8.99     6 sold      False   
24            narrowgate15         4934     US $11.99    15 sold      False   
25            narrowgate15         4934      US $9.99     5 sold      False   
26            narrowgate15         4934      US $9.99     2 sold       True   
27            narrowgate15         4934     US $11.99        NaN       True   
28            narrowgate15         4934      US $9.99     3 sold      False   
29            narrowgate15         4934     US $10.99        NaN       True   
30                be-start         8451  AU $3,211.55        NaN      False   
31                be-start         8451  AU $2,210.95        NaN      False   
32                be-start         8451  AU $2,210.95        NaN      False   
33                be-start         8451  AU $2,538.35        NaN      False   
34                be-start         8451  AU $2,538.35        NaN      False   
35                be-start         8451  AU $2,538.35        NaN      False   
36           summercloud14           48  US $1,490.00     2 sold       True   
37                be-start         8451  AU $3,211.55        NaN      False   
38          the_phone_king        14640  AU $2,437.99     1 sold      False   
39      everythingmustgo_8            1  US $1,199.00        NaN      False   
40                 2jspawn         2642  US $1,150.00        NaN       True   
41       usawoodbridge2010         5980    US $854.99        NaN      False   
42           adopt-a-phone         8650    US $924.99   303 sold       True   
43                  krunch           30  US $1,225.00        NaN       True   
44               antlam-11            4  US $1,300.00        NaN      False   
45             debillaslux          593  US $1,699.00        NaN       True   
46  advanced_power_systems          253  US $1,175.00        NaN      False   
47                jak28139          490  US $1,375.00        NaN       True   
48           happyversellc           14  US $1,489.00     2 sold      False   
49             ilnurgadget         6160  US $1,379.00     1 sold 

In [414]:
#QC:
#should be 59 (=61-2 as there are two duplicates among sponsored items - duplicates replaced 
#on creating file under same names)
len(scrape_df)

59

In [594]:
nonsponsored_df

seller_name seller_score               item_price  \
0                hammercosmetics          327             US $2,199.99   
1                hammercosmetics          327             US $2,199.99   
2                        dewdads         1194               US $690.00   
3                hammercosmetics          327             US $7,999.99   
4                     z3d_122324            2             US $1,600.00   
5                hammercosmetics          327             US $1,999.99   
6                      305racers           72             US $1,325.00   
7                 buytheworld999          890             US $1,379.00   
8                 buytheworld999          890               US $879.00   
9                      vipoutlet       258802        
							$1,199.00   
10                     vipoutlet       258803        
							$1,349.00   
11                     vipoutlet       258803        
							$1,049.00   
12                 the.lux.group          188             US $1,899.00   
13                     vipoutlet       258802        
							$1,149.99   
14                     vipoutlet       258802        
							$1,099.00   
15                     vipoutlet       258803        
							$1,249.00   
16                     vipoutlet       258800        
							$1,349.00   
17                     vipoutlet       258803        
							$1,349.00   
18                     vipoutlet       258800        
							$1,349.00   
19                     antik1513         2840             US $2,250.00   
20                 the.lux.group          188             US $1,899.00   
21                     vipoutlet       258802        
							$1,049.99   
22                     vipoutlet       258803        
							$1,149.00   
23                     vipoutlet       258802        
							$1,199.00   
24   mohammadrezarezaradmanesh90           40             US $1,079.00   
25                     vipoutlet       258803        
							$1,049.99   
26                 the.lux.group          188             US $1,999.00   
27                     vipoutlet       258803        
							$1,349.00   
28                     vipoutlet       258803        
							$1,349.00   
29                        ecoatm        11490     
							US $1,099.00   
..                           ...          ...                      ...   
964         trinityconnectonline         2007             AU $2,422.15   
965         trinityconnectonline         2007             AU $2,422.15   
966         trinityconnectonline         2007             AU $2,538.35   
967         trinityconnectonline         2007             AU $2,538.35   
968                   limetropic        78243               US $866.28   
969         trinityconnectonline         2007             AU $2,210.95   
970                    tabgre-73           10             US $1,475.00   
971                   migdu_8308           34               US $619.99   
972                 youngs_stuff           83             US $1,099.99   
973                 youngs_stuff           83  
										US $1,399.99   
974                     jentr_76           15               US $700.00   
975                    pawa-2997            0    
										US $899.00   
976               onlinemart1111          113               US $499.99   
977                    emwa-2988           16             US $1,099.99   
978                  koupacome_0           21               US $749.00   
979               onlinemart1111          113               US $599.77   
980                        onha1         3968               US $900.00   
981                   mr.goodkat          621               US $499.94   
982           mainly-electronics        70717             US $1,249.99   
983            wi_zacza_nakykqb3            2    
										US $991.00   
984                    hilapr313          191                US $13.89   
985              muchmoreforless        15490             US $1,749.98   
986              muchmore

In [593]:
#QC-6 are duplicated 
len(nonsponsored_df)

994

In [996]:
#Appending both the dataframes into one for better visual check
interim_df=sponsored_df.append(nonsponsored_df, ignore_index=True)

In [997]:
interim_df

seller_name seller_score               item_price units_sold  \
0            the.lux.group          188             US $2,349.00        NaN   
1            the.lux.group          188             US $1,799.00     1 sold   
2            the.lux.group          188             US $1,999.00     1 sold   
3             normabreede0           15             US $1,500.00        NaN   
4              klp_customs          148             US $1,799.00        NaN   
5              klp_customs          148             US $1,799.00     1 sold   
6                  merinsh           84             US $1,700.00     2 sold   
7                  dheader        75647               US $699.95        NaN   
8       harvardhoney4545_1          184             US $1,699.99        NaN   
9                shaneel02           99             US $2,499.99     1 sold   
10         swiftphonetechs          127               US $859.00        NaN   
11                punkydk0          589               US $719.99        NaN   
12                punkydk0          589               US $569.99        NaN   
13         swiftphonetechs          127               US $925.00        NaN   
14            vombotsupply         5881             US $1,029.99   150 sold   
15          the_phone_king        14640             AU $3,047.99        NaN   
16          the_phone_king        14640             AU $2,315.99        NaN   
17          the_phone_king        14640             AU $2,132.99        NaN   
18                be-start         8451             AU $2,210.95        NaN   
19                be-start         8451             AU $3,211.55     1 sold   
20                be-start         8451             AU $3,029.35        NaN   
21            narrowgate15         4934                US $10.99     8 sold   
22            narrowgate15         4934                 US $8.99     2 sold   
23            narrowgate15         4934                 US $8.99     6 sold   
24            narrowgate15         4934                US $11.99    15 sold   
25            narrowgate15         4934                 US $9.99     5 sold   
26            narrowgate15         4934                 US $9.99     2 sold   
27            narrowgate15         4934                US $11.99        NaN   
28            narrowgate15         4934                 US $9.99     3 sold   
29            narrowgate15         4934                US $10.99        NaN   
...                    ...          ...                      ...        ...   
1023  trinityconnectonline         2007             AU $2,422.15        NaN   
1024  trinityconnectonline         2007             AU $2,422.15        NaN   
1025  trinityconnectonline         2007             AU $2,538.35        NaN   
1026  trinityconnectonline         2007             AU $2,538.35        NaN   
1027            limetropic        78243               US $866.28     6 sold   
1028  trinityconnectonline         2007             AU $2,210.95        NaN   
1029             tabgre-73           10             US $1,475.00        NaN   
1030            migdu_8308           34               US $619.99        NaN   
1031          youngs_stuff           83             US $1,099.99        NaN   
1032          youngs_stuff           83  
										US $1,399.99        NaN   
1033              jentr_76           15               US $700.00        NaN   
1034             pawa-2997            0    
										US $899.00        NaN   
1035        onlinemart1111          113               US $499.99        NaN   
1036             emwa-2988           16             US $1,099.99        NaN   
1037           koupacome_0           21               US $749.00        NaN   
1038        onlinemart1111          113               US $599.77        NaN   
1039                 onha1         3968               US $900.00     3 sold   
1040            mr.goodkat          621               US $499.94        NaN   
1041    mainly-electronics        70717             US $1,249.99 

<h3>2.e</h3><br>
<h5>Use your code script to connect to SQL (either MySQL, MariaDB, or SQLite. Do NOT use SQL GUI or command terminal). Create a database and name it "eBay". Save the information of items in (d)  into a single table named "eBay_items"</h5><br>
(You are allowed to use only one table).


Cleaning the dataset properly<br>
<li>Changing price from US dollars to cent format. Prices to be inserted as INT
<li>Changing Yes/No or True/false columns into binary (1/0) format
<li>Treating for missing values and Nan values in the dataset should be changed to NULL

<h5>MySQL Python Connector</h5>
    1.Setting up the SQL connection<br>
    2.Creating the database, <b>eBay</b><br>
    3.Creating the MySQL table, <b>eBay_items</b><br>

In [985]:
#Connecting to MySQL using Python connector
import mysql.connector

#establishing the connection
#entering the database names as ucdavis is itsn't there it gets created automatically
conn = mysql.connector.connect(
   user='root', password='', host='localhost'
)

#Creating a cursor object using the cursor() method
cursor = conn.cursor()

#Executing cursor with execute method and pass SQL query
cursor.execute("CREATE DATABASE eBay")

#Get list of all databases to check
cursor.execute("SHOW DATABASES")
#print all databases
for db in cursor:
    print(db)

cursor.execute("USE eBay")
#Dropping table if already exists.
cursor.execute("DROP TABLE IF EXISTS eBay.eBay_items")

#Creating table as per requirement
sql ='''CREATE TABLE eBay_items(
   itemID INT(4) PRIMARY KEY AUTO_INCREMENT,
   item_title VARCHAR(1000),
   item_price INT(5),
   seller_name VARCHAR(255),
   seller_score INT(5), 
   units_sold INT(5), 
   returns_allowed INT(1), 
   best_offer INT(1), 
   shipping_price INT(5),
   shipping_zip INT(1),
   item_condition VARCHAR(100), 
   sponsored INT
)'''

cursor.execute(sql)

#Checking where the cursor is at
for table in cursor:
    print(table)
    

('accountspayable',)
('avidreader',)
('bowlingleagueexample',)
('colonial',)
('ebay',)
('entertainmentagencyexample',)
('information_schema',)
('mysql',)
('oact',)
('performance_schema',)
('pythondatabase',)
('recipesdatabase',)
('sakila',)
('salesordersdatabase',)
('sys',)
('ucdavis',)
('world',)


### Best Offer
Convert such that 0- Best offer option unavailable, 1- Best offer option available
### Units Sold
Remove 'sold' from the values using regex and wherever the info unavailable) store it as <b>None</b>
### Item Price
Using Regex, cleaned the entries<br>
Stripped Currency type using search for alphabets and similarly value using search<br>
Wherever no info on currency type was given, I assumed it to be US dollars.<br>
Applied necessary conversion rate on the value<br>
Multiplied it by 100<br>
Converted the data type to int<br>
Applying this conversion factors, modifying all prices to USD<br>
1 GBP=1.29 USD<br>
1 CAD=0.75 USD<br>
1 AUD=0.67 USD<br>

### Shipping Price
Every shipping price in other currency have their corresponding US dollars price which has been retained dropping everything else using regex. Similarly, newline and tab characters are removed
Also, please note <b>'Local pickup offered'/'Free Local Pickup'</b> has been assumed as <b>0</b> Or where it needs to be computed based on shipping <b>Calculate</b> as <b>0</b><br>
I have created a separate column called <b>"shipping_zip"</b> to differentiate between those with fixed rates(value=0) and those with rates depending on zip codes(value=1)

### Returns Allowed
We have information for how many days returns are allowed, for how long it's free if available etc.<br>
But to make it simpler,returns_allowed will be a binary variable indicating if returns are allowed or not<br>
0-not allowed, 1-allowed

In [986]:
print(cursor)

CMySQLCursor: CREATE TABLE eBay_items(
   itemID INT(4..


#### Created a Function that scrapes the files in the directory and adds to the MySQL table created above
Passing Parameters: path to the directory, value for sponsor(1:sponsored or 0:non-sponsored), cursor(mySQL cursor)


In [987]:
def scrape_storedb(path,sponsor_flag,cursor):
    #for each file in the directory getting the data  and inserting into the sql table eBay_items
    for file in os.listdir(path):
        #open each file in read bytes mode
        f=open(file, "rb")
        filetext=f.read()
        #parse it through Beautiful soup 
        data = BeautifulSoup(filetext, 'html.parser')

        #Get Seller-name
        seller_name=None if data.find('span',class_='mbg-nw') is None else data.find('span',class_='mbg-nw').string

        #store seller score in integer format
        seller_score=np.nan if data.select('span.mbg-l > a') is None else int(data.select('span.mbg-l > a')[0].string)

        #ITEM PRICE
        item_price= np.nan if data.find('span',class_='notranslate')is None else data.find('span',class_='notranslate').string
        #,itemprop='price'

        item_price = re.sub('\n+|\t+|,|\$|\/ea','',item_price)
        #split the column into currency type  and amount
        item_c= None if (re.search('[A-Z]+',item_price)) is None else (re.search('[A-Z]+',item_price)).group(0)
        item_price=(re.search('[0-9]+.[0-9]+',item_price)).group(0)
        
        if item_c=='US':
            currency=1
        elif item_c=='AU':
            currency=0.67
        elif item_c=='C':
            currency=0.75
        elif item_c=='GBP':
            currency=1.25
        else:currency=1 #Cases where no country is mentioned, default as US

        #Changing the data type
        item_price=float(item_price)*currency*100
        item_price=int(item_price)

        #UNITS SOLD
        #if NoneType object is found then store as nan(value unavailable)
        units_sold= None if data.find('a',class_='vi-txt-underline') is None else data.find('a',class_='vi-txt-underline').string
        #Regex for sold substring from the quantities
        units_sold= units_sold if units_sold is None else re.sub(' sold','',units_sold)

        #BEST OFFER AVAILABLE
        #if we get NoneType object, best offer doesn't exist for the item
        best_offer= True if data.find('div',class_=re.compile('.*boLable')) is not None else False
        #Best Offers as 0/1
        best_offer= int(1 if best_offer==True else 0)

        #Item title
        title1=data.find('h1',class_='it-ttl',id='itemTitle').get_text()
        #removing unecessary part of the string
        item_title=re.sub('.*\xa0', '', title1)

        #RETURNS DETAILS
        returns_allowed= None if data.find('span',id='vi-ret-accrd-txt') is None else data.find('span',id='vi-ret-accrd-txt').string
        #Returns Allowed-0/1
        returns_allowed= int (0 if returns_allowed =='Seller does not accept returns' else 1)

        #SHIPPING PRICE
        #finding shipping summary then looking for next as its dynamic across product pages
        shipping= data.find('span',id='shSummary')
        shipping_price=shipping.find_next().get_text()

        #converting into integers
        shipping_price = re.sub('\n+|\t+','',shipping_price)
        shipping_price = re.sub('(AU|GBP|C)( ){1,2}\${0,1}[0-9]+.[0-9]+\(approx.( ){0,2}US|\)|\$','',shipping_price)

        #if the shipping is variable 1 else 0
        shipping_zip= 1 if shipping_price in('Local pick-up offered.','Calculate','Free Local Pickup') else 0

        #All free shipping have the value 0
        def shipping(x):
            if x =='FREE': 
                return 0
            elif x =='FAST \'N FREE': 
                return 0
            elif x=='Calculate': 
                return 0
            elif x=='Local pick-up offered.': 
                return 0
            elif x=='Free Local Pickup':
                return 0
            else: return  x

        #Changing data-type
        shipping_price=shipping(shipping_price)
        shipping_price= float(shipping_price)
        #Converting to INT
        shipping_price= int(shipping_price*100)

        #Item Condition
        item_condition= None if data.find('div',id='vi-itm-cond',itemprop='itemCondition') is None else data.find('div',id='vi-itm-cond',itemprop='itemCondition').string

        #sponsor flag
        sponsored=sponsor_flag

        #Close the file
        f.close()

        #inserting into the table omdb
        sql_query = "INSERT INTO eBay_items(item_title, item_price, seller_name,seller_score,units_sold,returns_allowed,best_offer,shipping_price,shipping_zip,item_condition,sponsored) VALUES(%s, %s, %s, %s, %s, %s,%s, %s, %s, %s, %s)"
        record=(item_title, item_price, seller_name,seller_score,
                units_sold,returns_allowed,best_offer,shipping_price,
                shipping_zip,item_condition,sponsored)
        cursor.execute(sql_query,record)
        #print("inserted")
        #print("\n")

    

First inserting sponsored items<br>
1.Entering into Sponsored Folder<br>
2.Traversing through each of the files using a function created above<br>
3.Passing sponsor-flag as 1 for sponsored items<br>

In [989]:
#Getting outside of Non-Sponsored folder
os.chdir('../')
#Reset the directory
os.getcwd()

'C:\\Users\\shali\\OneDrive\\Documents\\3 WQ Data Design\\Assignments Submitted'

In [990]:
org_path=os.getcwd()
new_path=org_path+"\Sponsored"
#Change the current working direcory: 
os.chdir(new_path)
#sponsor-flag=1 for sponsored items
scrape_storedb(new_path,1,cursor)

inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted




Second, inserting Non-sponsored items<br>
1.Exiting Sponsored Folder<br>
2.Entering Non-sponsored Folder<br>
3.Traversing through each of the files using a function created above<br>
4.Passing sponsor-flag as 0 in the function for non-sponsored items<br>

In [991]:
conn.commit()
#Getting outside of Sponsored folder
os.chdir('../')
#Reset the directory
os.getcwd()

'C:\\Users\\shali\\OneDrive\\Documents\\3 WQ Data Design\\Assignments Submitted'

In [992]:
org_path=os.getcwd()
new_path=org_path+"\/"+"Non-Sponsored"
#Change the current working direcory: 
os.chdir(new_path)
#sponsor-flag=0 for non-sponsored items
scrape_storedb(new_path,0,cursor)

inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted



inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted


inserted



In [993]:
conn.commit()
#validating whether records were inserted
cursor.execute("select * from eBay.eBay_items")
records = cursor.fetchall()
print("Total number of rows in eBay_items is: ", cursor.rowcount)
#Getting back to the main folder
os.chdir('../')

Total number of rows in eBay_items is:  1053


In [994]:
#Closing the connection
cursor.close()
conn.close()

### 2.f
<h5>Use your code script (and NOT SQL GUI or command terminal) to run summary stats on each item. Print to the screen the mean, min, max, and mean for each column, grouped by "sponsor/non-sponsor" and "condition" (group by at the same time, not separately).</h5>
For binary categorical columns, use 0-1 conversion.<br>
For e.g., for the "returns allowed" convert YES to 1 and NO to 0 and then calculate the stats. If it is NOT a numerical/binary categorical column, print to the screen the count of each category level.<br>
You will need to ignore NULL values in your statistic calculations. 

<h4>Answer:</h4>
Aggregate (summary) functions such as COUNT(), MIN(), and SUM() ignore NULL values. The exception to this is COUNT(*), which counts rows and not individual column values.
Reference:<u>https://dev.mysql.com/doc/refman/8.0/en/problems-with-null.html</u><br><br>
For Averages: I am using the formula SUM(records)/COUNT(records) which will ignore NULLs as each of these functions ignore NULLs<br>
We have NULLs in the column <b>units sold</b> because for few items it was not given<br>
After that also,<br> we see some null values because they are just NULL records for that particular row(sponsored,condition) combination in our entire table.

In [998]:
#Connecting to MySQL using Python connector
import mysql.connector

#establishing the connection
#Adding an exception block to know connection status

try:
  conn = conn = mysql.connector.connect(
   user='root', password='', host='localhost',database='eBay'
)
except mysql.connector.Error as err:
  if err.errno == errorcode.ER_ACCESS_DENIED_ERROR:
    print("Something is wrong with your user name or password")
  elif err.errno == errorcode.ER_BAD_DB_ERROR:
    print("Database does not exist")
  else:
    print(err)
else:
   if (conn.is_connected()):
        print("Connected")
        
cursor = conn.cursor()
#cursor.execute("DROP TABLE IF EXISTS ucdavis.OMDB")

Connected


In [999]:
#Running query to get statistics as asked in the question
cursor.execute("SELECT sponsored,item_condition,count(itemID) as item_count,min(seller_score) as min_seller_score,max(seller_score) as max_seller_score,avg(seller_score) as avg_seller_score,stddev(seller_score) as sdev_seller_score,min(item_price) as min_price,max(item_price) as max_price,avg(item_price) as avg_price,stddev(item_price) as sdev_price,min(shipping_price) as min_ship_price,max(shipping_price) as max_ship_price,avg(shipping_price) as avg_ship_price,stddev(shipping_price) as sdev_ship_price,min(returns_allowed) as min_returns,max(returns_allowed) as max_returns,avg(returns_allowed) as avg_returns,stddev(returns_allowed) as sdev_returns,min(best_offer) as min_b_offers,max(best_offer) as max_b_offers,avg(best_offer) as avg_b_offers,stddev(best_offer) as sdev_b_offer,min(units_sold) as min_units,max(units_sold) as max_units,SUM(units_sold)/COUNT(units_sold) as avg_units,stddev(units_sold) as sdev_units FROM eBay_items GROUP BY sponsored,item_condition;")
records = cursor.fetchall()
print('Total Row(s):', cursor.rowcount)
for record in records:
    print(record)

Total Row(s): 12
(1, 'Open box', 8, 1, 8650, Decimal('2972.0000'), 3769.5460933645577, 84999, 234900, Decimal('142199.6250'), 51850.88258876964, 0, 0, Decimal('0.0000'), 0.0, 0, 1, Decimal('0.3750'), 0.4841229182759271, 0, 1, Decimal('0.7500'), 0.4330127018922193, 1, 303, Decimal('98.0000'), 116.2032701777364)
(1, 'New', 39, 4, 60475, Decimal('8132.5897'), 12973.923321205122, 899, 510000, Decimal('144589.3590'), 113336.81948714143, 0, 20233, Decimal('3519.6154'), 6022.680290179118, 0, 1, Decimal('0.3590'), 0.4796996649710181, 0, 1, Decimal('0.2564'), 0.4366509324596513, 1, 18, Decimal('4.1667'), 4.763868642652151)
(1, 'For parts or not working', 7, 127, 75647, Decimal('11787.5714'), 26091.438814058438, 56999, 115000, Decimal('83484.5714'), 17599.522565416715, 0, 700, Decimal('100.0000'), 244.9489742783178, 0, 1, Decimal('0.2857'), 0.45175395145262565, 0, 1, Decimal('0.1429'), 0.3499271061118826, 1, 1, Decimal('1.0000'), 0.0)
(1, 'Used', 2, 3588, 5980, Decimal('4784.0000'), 1196.0, 8549

In [1017]:
#Viewing the results in tabular format
results_table=pd.DataFrame(records)
results_table.columns=("sponsored","item_condition","item_count","min_seller_score","max_seller_score","avg_seller_score","sdev_seller_score","min_price","max_price","avg_price","sdev_price","min_ship_price","max_ship_price","avg_ship_price","sdev_ship_price","min_returns","max_returns","avg_returns","sdev_returns","min_b_offers","max_b_offers","avg_b_offers","sdev_b_offers","min_units","max_units","avg_units","sdev_units")
results_table


sponsored            item_condition  item_count  min_seller_score  \
0           1                  Open box           8                 1   
1           1                       New          39                 4   
2           1  For parts or not working           7               127   
3           1                      Used           2              3588   
4           1        Seller refurbished           2               253   
5           1  Manufacturer refurbished           1               299   
6           0                       New         651                 0   
7           0  For parts or not working         117                 3   
8           0                  Open box         113                 0   
9           0        Seller refurbished          44                68   
10          0                      Used          66                 0   
11          0  Manufacturer refurbished           3              1512   

    max_seller_score avg_seller_score  sdev_seller_score  min_price  \
0               8650        2972.0000       3.769546e+03      84999   
1              60475        8132.5897       1.297392e+04        899   
2              75647       11787.5714       2.609144e+04      56999   
3               5980        4784.0000       1.196000e+03      85499   
4                299         276.0000       2.300000e+01     117500   
5                299         299.0000       0.000000e+00     104000   
6             258803        3394.0476       1.229401e+04        509   
7              46066        4477.3675       8.921172e+03      36800   
8             325998       14550.7434       4.398804e+04        600   
9            2265053      279374.6818       4.467676e+05      86628   
10            258803       26740.6061       6.124701e+04      50000   
11           2265053      836526.6667       1.014921e+06      84999   

    max_price    avg_price  ...  avg_returns  sdev_returns  min_b_offers  \
0      234900  142199.6250  ...       0.3750      0.484123             0   
1      510000  144589.3590  ...       0.3590      0.479700             0   
2      115000   83484.5714  ...       0.2857      0.451754             0   
3      109900   97699.5000  ...       1.0000      0.000000             0   
4      117500  117500.0000  ...       0.5000      0.500000             0   
5      104000  104000.0000  ...       1.0000      0.000000             0   
6      799999   82746.9892  ...       0.7358      0.440911             0   
7      119900   74855.7094  ...       0.4103      0.491880             0   
8      214900  108728.7345  ...       0.6726      0.469277             0   
9      144900  115388.4545  ...       0.9545      0.208299             0   
10     199900  100411.8030  ...       0.6515      0.476490             0   
11      91995   88997.6667  ...       0.6667      0.471405             0   

   max_b_offers  avg_b_offers  sdev_b_offers  min_units max_units  avg_units  \
0             1        0.7500       0.433013        1.0     303.0    98.0000   
1             1        0.2564       0.436651        1.0      18.0     4.1667   
2             1        0.1429       0.349927        1.0       1.0     1.0000   
3             0        0.0000       0.000000        NaN       NaN       None   
4             0        0.0000       0.000000        1.0       1.0     1.0000   
5             0        0.0000       0.000000        NaN       NaN       None   
6             1        0.1029       0.303853        1.0     359.0     7.8333   
7             1        0.2564       0.436651        1.0      11.0     3.5000   
8             1        0.2655       0.441592        1.0      30.0     6.7273   
9             0        0.0000       0.000000        1.0      29.0     6.8000   
10            1        0.1364       0.343174        1.0       3.0     1.3333   
11            0        0.0000       0.000000      110.0     110.0   110.0000   

    sdev_units  
0   116.203270  
1     4.763869  
2     0.000000  
3          NaN  
4     0.000000 

In [1019]:
cursor.close()
conn.close()

### 2.g
<h5>Use the stats in (f) and tell in your own words, how sponsored and non-sponsored items appear to be different. Is there a variable/column that can be used to predict the sponsor/non-sponsor items.</h5><br> (You do NOT need to run any model or statistic calculations for this part. Make your judgment only by looking at the stats you printed in (f))

The distribution of items based on their condition is kind of consistent across sponsored/non-sponsored group i.e.66% new items in sponsored group whereas 65.4% in non-sponsored group. similarly for 'parts not working' ~ 12% in both. But the behavior of different variables in interaction with condition is very differeing. Item condition can be used as a viable interaction variable in our model.
Shipping price for non-sponsored items are in general priced higher than sponsored items. For e.g. for open box items shipping price is on an average 2.2 dollars whereas it's free in case of sponsored items. Reduced or free shipping seems to be one of the glam factor for sponsored items. this difference in their behavior will make it a good regressor to be considered for our model.
New sponsored items are priced higher than the non-sponsored items. But as we move from best to reduced condition, the price tends to get more cheaper for sponsored compared to non-sponsored, meaning chances of getting a cheaper refurbished sponsored item is higher than that of a non sponsored one. Item Price can also be used as a regressor.
More return options are provided by sellers when selling a non-sponsored items. For sponsored items of conditions other than new or used or for parts, there are no returns available.Indicative of the fact that returns can be used as a regressor too.
Also buyer offers are higher in case of new sponsored items and also in sponsored items in overall. We can use it with item condition as an interaction variable
For units sold, units sold for new sponsored items on an average is exceedingly high whereas among non-sponsored and other sponsored items there isn't much of a change.But within new sponsored items also there is a wide variation with units ranging from 1 to 303. New non-sponsored items also have a maximum units sold of 359, a number higher than sponsored but with much less standard deviation.On an overall basis, average units sold ia higher for non-sponsored. We should try running our model excluding units as it introduces more variation rather than direction in the split.But we should first try without it.